libweb3jsonrpc/Debug.cpp

```diff
@@ -126,6 +126,38 @@ Json::Value Debug::debug_traceBlockByNumber(int _blockNumber, Json::Value const&
 	return ret;
 }
 
+Json::Value Debug::debug_accountRangeAt(string const& _blockHashOrNumber, int _txIndex, string const& _address, int _maxResults)
+{
+	Json::Value ret(Json::arrayValue);
+
+	if (_txIndex < 0)
+		throw jsonrpc::JsonRpcException("Negative index");
+	if (_maxResults <= 0)
+		throw jsonrpc::JsonRpcException("Nonpositive maxResults");
+
+	try
+	{
+		Block block = m_eth.block(blockHash(_blockHashOrNumber));
+
+		unsigned const i = ((unsigned)_txIndex < block.pending().size()) ? (unsigned)_txIndex : block.pending().size();
+		State state(State::Null);
+		createIntermediateState(state, block, i, m_eth.blockChain());
+		Address from(_address);
+		for (auto const& addr : state.addresses())
```

Comment:
so the better way I see would be to sort addresses first then just select address

Reply:
Not at all. The problem is you have to load all accounts from database first. That's not a big deal in test chain but on the mainnet that's not feasible. You have to extend State interface to support returning a subset of addresses.

Reply:
Also it should return hashes of addresses, not addresses, because addresses are not always available (e.g. not available after snapshot sync)

@cdetrio alsp mentioned it in https://github.com/ethereum/retesteth/issues/5#issuecomment-383314903

Reply:
hmm. not sure what I would do with address hashes. if this function return list of hashes I wont be able to tell which address is unexpectedly exist in the post state.

Reply:
Why not? You'd compare hashes of expected addresses with hashes returned from here. You'd find the hash of unexpected address

Reply:
I think there should be 2 options. one return only hashes and another one return addresses.
If I know the adresses I could tell which hash coresspond which adress. 
But if I don't know the address I have no means to restore the address from it's hash. (e.x a  new address that is being created)

Reply:
BTW keys returned from `debug_storageRangeAt` may be empty, too, for the same reason.
Only hashes of storage keys are always returned.

Reply:
The adresses method just iterate over m_state and m_cache. Where the loading from db process takes place?

The hashes are not acceptable. Post state requires adresses not hashes.

Reply:
`m_state` is a `SecureTrieDB` which will acess DB when you iterate over the trie.

> The hashes are not acceptable. Post state requires adresses not hashes.
Then it will work only with archive nodes.
I still would prefer this method to be general and return both address and hashed address, where the address is optional (similar to what `debug_storageRangeAt` returns)

Reply:
Sure it's ordered, it's a trie. I think you should utilize `lower_bound` method of the Trie classes. Somewhat similar to how `debug_storageRangeAt` finds the point to start iterating.

Reply:
@gumb0 The key in the database is the hash of the address? so we cannot filter the addresses without first loading it into the cache?

Reply:
That was my intention here from the beginning.